crates/wasmtime/src/runtime/component/func/host.rs

```diff
@@ -270,21 +361,65 @@ where
             &mut self,
             cx: &mut LowerContext<'_, T>,
             ty: InterfaceType,
-            ret: R,
+            ret: ReturnMode<R>,
```

Comment:
Just a hunch but after reading through this logic, I wonder: would it clean up the flow somewhat if we dispatched once on `ReturnMode` at the top level, then avoided all of the separate conditionals throughout the body? As-is, it's quite difficult to follow the different paths -- it seems somewhat decoupled and overly-abstracted.